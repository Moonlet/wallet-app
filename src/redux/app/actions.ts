import { HintsComponent, HintsScreen } from './state';
import { IReduxState } from '../state';
import { clearPinCode } from '../../core/secure/keychain';

export const SET_ACCEPTED_TC_VERSION = 'SET_ACCEPTED_TC_VERSION';
export const SHOW_HINT = 'SHOW_HINT';
export const RESET_FAILED_LOGINS = 'RESET_FAILED_LOGINS';
export const INCREMENT_FAILED_LOGINS = 'INCREMENT_FAILED_LOGINS';
export const SET_APP_BLOCK_UNTIL = 'SET_APP_BLOCK_UNTIL';
export const RESET_ALL_DATA = 'RESET_ALL_DATA';

export const appSetAcceptedTcVersion = (version: number) => {
    return {
        type: SET_ACCEPTED_TC_VERSION,
        data: version
    };
};

export const updateDisplayedHint = (screen: HintsScreen, component: HintsComponent) => {
    return {
        type: SHOW_HINT,
        data: { screen, component }
    };
};

export const incrementFailedLogins = () => {
    return {
        type: INCREMENT_FAILED_LOGINS
    };
};

export const resetFailedLogins = () => {
    return {
        type: RESET_FAILED_LOGINS
    };
};

export const setAppBlockUntil = (date: Date | string) => {
    return {
        type: SET_APP_BLOCK_UNTIL,
        data: { date }
    };
};

export const resetAllData = () => async (dispatch, getState: () => IReduxState) => {
    clearPinCode(); // clear keychain storage
    dispatch({ type: RESET_ALL_DATA });
};
