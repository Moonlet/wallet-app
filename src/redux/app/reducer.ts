import { IAppState } from './state';
import { IAction } from '../types';
import {
    SET_ACCEPTED_TC_VERSION,
    SHOW_HINT,
    RESET_FAILED_LOGINS,
    INCREMENT_FAILED_LOGINS,
    SET_APP_BLOCK_UNTIL,
    RESET_ALL_DATA,
    REDUX_UPDATE_STATE
} from './actions';

const intialState: IAppState = {
    version: 1,
    tcAcceptedVersion: undefined,
    hints: {
        SEND_SCREEN: {
            ADDRESS_BOOK: 0
        },
        WALLETS_SCREEN: {
            WALLETS_LIST: 0
        },
        MANAGE_ACCOUNT: {
            TOKENS_LIST: 0
        }
    },
    failedLogins: 0,
    blockUntil: undefined
};

export default (state: IAppState = intialState, action: IAction): IAppState => {
    switch (action.type) {
        case SET_ACCEPTED_TC_VERSION:
            return {
                ...state,
                tcAcceptedVersion: action.data
            };

        case SHOW_HINT:
            return {
                ...state,
                hints: {
                    ...state.hints,
                    [action.data.screen]: {
                        [action.data.component]:
                            state.hints[action.data.screen][action.data.component] + 1
                    }
                }
            };

        case RESET_FAILED_LOGINS:
            return {
                ...state,
                failedLogins: 0
            };

        case INCREMENT_FAILED_LOGINS:
            return {
                ...state,
                failedLogins: state.failedLogins + 1
            };

        case SET_APP_BLOCK_UNTIL:
            return {
                ...state,
                blockUntil: action.data.date
            };

        case RESET_ALL_DATA:
            return intialState;

        case REDUX_UPDATE_STATE: {
            return {
                ...state,
                version: action.data.state.app.version
            };
        }

        default:
            break;
    }
    return state;
};
