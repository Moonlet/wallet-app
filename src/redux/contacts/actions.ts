import { IContactState } from './state';

// actions consts
export const CONTACT_ADD = 'CONTACT_ADD';

export const addContact = (contactData: IContactState) => {
    return {
        type: CONTACT_ADD,
        data: contactData
    };
};
