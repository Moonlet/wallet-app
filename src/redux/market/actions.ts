export const EXCHANGE_RATE_UPDATE = 'EXCHANGE_RATE_UPDATE';

export const updateExchangeRate = (exchangeRate: { token: string; value: string }) => {
    return {
        type: EXCHANGE_RATE_UPDATE,
        data: exchangeRate
    };
};
