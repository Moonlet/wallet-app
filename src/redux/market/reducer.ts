import { IAction } from '../types';

const initialState = {
    usdPrices: {
        USD: 1,
        BTC: 8349.75,
        ETH: 182.25,
        ZIL: 0.005821,
        ATOM: 3.03,
        XLM: 0.06122
    },
    change: {
        daily: {
            USD: {
                USD: 0,
                BTC: 0.0045,
                ETH: -0.0016
            },
            BTC: {
                USD: 0.0043,
                BTC: 0,
                ETH: -0.0016
            },
            ETH: {
                USD: 0.0143,
                BTC: 0.0045,
                ETH: 0
            },
            ZIL: {
                USD: 0.0002,
                BTC: 0.0045,
                ETH: 0.0216
            },
            ATOM: {
                USD: 0.0045,
                BTC: -0.0045,
                ETH: -0.0016
            },
            XLM: {
                USD: 0.0045,
                BTC: 0.0033,
                ETH: -0.0016
            }
        }
    }
};

const intialState: any = initialState;

export default (state: any = intialState, action: IAction): any => {
    return state;
};
