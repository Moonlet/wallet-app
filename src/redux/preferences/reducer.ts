import { IPrefState } from './state';
import { IAction } from '../types';
import {
    PREF_SET_CURRENCY,
    TOGGLE_TOUCH_ID,
    SET_TEST_NET,
    PREF_SET_NETWORK_TEST_NET_CHAIN_ID,
    PREF_SET_BLOCKCHAIN_ACTIVE_STATE,
    PREF_SET_BLOCKCHAIN_ORDER
} from './actions';
import { Blockchain } from '../../core/blockchain/types';

const initialState: IPrefState = {
    currency: 'USD',
    testNet: false,
    touchID: false,
    networks: {
        [Blockchain.ZILLIQA]: {
            testNet: 333,
            mainNet: 1
        },
        [Blockchain.ETHEREUM]: {
            testNet: 4,
            mainNet: 1
        },
        [Blockchain.NEAR]: {
            testNet: 'testnet'
        }
    },
    blockchains: {
        [Blockchain.ZILLIQA]: {
            order: 0,
            active: true
        },
        [Blockchain.ETHEREUM]: {
            order: 1,
            active: true
        },
        [Blockchain.NEAR]: {
            order: 2,
            active: true
        }
    }
};

export default (state: IPrefState = initialState, action: IAction): IPrefState => {
    switch (action.type) {
        case PREF_SET_NETWORK_TEST_NET_CHAIN_ID:
            return {
                ...state,
                networks: {
                    ...state.networks,
                    [action.data.blockchain]: {
                        ...state.networks[action.data.blockchain],
                        testNet: action.data.chainId
                    }
                }
            };
        case PREF_SET_BLOCKCHAIN_ACTIVE_STATE:
            return {
                ...state,
                blockchains: {
                    ...state.blockchains,
                    [action.data.blockchain]: {
                        ...state.blockchains[action.data.blockchain],
                        active: action.data.active
                    }
                }
            };
        case PREF_SET_BLOCKCHAIN_ORDER:
            return {
                ...state,
                blockchains: action.data.blockchains
            };
        case SET_TEST_NET:
            return {
                ...state,
                testNet: !state.testNet
            };
        case TOGGLE_TOUCH_ID:
            state = { ...state };
            state.touchID = !state.touchID;
            break;
        case PREF_SET_CURRENCY:
            return {
                ...state,
                currency: action.data.currency
            };
        default:
            break;
    }
    return state;
};
