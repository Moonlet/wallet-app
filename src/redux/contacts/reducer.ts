import { IAction } from '../types';
import { IContactsState } from './state';
import { CONTACT_ADD, CONTACT_DELETE, CONTACT_UPDATE_NAME } from './actions';

const intialState: IContactsState = {};

export default (state: IContactsState = intialState, action: IAction) => {
    switch (action.type) {
        case CONTACT_ADD:
            return {
                ...state,
                [action.data.address]: action.data
            };

        case CONTACT_DELETE:
            return Object.keys(state)
                .filter(key => key !== action.data.address)
                .reduce((result, current) => {
                    result[current] = state[current];
                    return result;
                }, {});

        case CONTACT_UPDATE_NAME:
            const newState = state;
            newState[action.data.address] = action.data;
            return newState;

        default:
            break;
    }
    return state;
};
