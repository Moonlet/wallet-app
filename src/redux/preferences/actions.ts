import { Dispatch } from 'react';
import { IReduxState } from '../state';
import { getSelectedAccount, getSelectedBlockchain } from '../wallets/selectors';
import { getBalance, setSelectedBlockchain } from '../wallets/actions';
import { Blockchain, ChainIdType } from '../../core/blockchain/types';
import { IBlockchainsOptions } from './state';
import { IAction } from '../types';
import { getBlockchains, hasNetwork } from './selectors';

// actions consts
export const PREF_SET_CURRENCY = 'PREF_SET_CURRENCY';
export const TOGGLE_TOUCH_ID = 'TOGGLE_TOUCH_ID';
export const SET_TEST_NET = 'SET_TEST_NET';
export const PREF_SET_BLOCKCHAIN_ACTIVE_STATE = 'PREF_SET_BLOCKCHAIN_ACTIVE_STATE';
export const PREF_SET_BLOCKCHAIN_ORDER = 'PREF_SET_BLOCKCHAIN_ORDER';
export const PREF_SET_NETWORK_TEST_NET_CHAIN_ID = 'PREF_SET_NETWORK_TEST_NET_CHAIN_ID';

export const setBlockchainActive = (blockchain: Blockchain, active: boolean) => (
    dispatch: Dispatch<IAction>,
    getState: () => IReduxState
) => {
    const state = getState();
    const selectedBlockchain = getSelectedBlockchain(state);

    if (selectedBlockchain === blockchain && active === false) {
        const nextBlockchain = Object.keys(state.preferences.blockchains).find(
            object => object !== blockchain
        );
        if (nextBlockchain) {
            setSelectedBlockchain(Blockchain[nextBlockchain])(dispatch, getState);
        }
    }
    dispatch({
        type: PREF_SET_BLOCKCHAIN_ACTIVE_STATE,
        data: { blockchain, active }
    });
};

export const setBlockchainOrder = (blockchains: IBlockchainsOptions[]) => {
    return {
        type: PREF_SET_BLOCKCHAIN_ORDER,
        data: { blockchains }
    };
};

export const setNetworkTestNetChainId = (blockchain: Blockchain, chainId: ChainIdType) => {
    return {
        type: PREF_SET_NETWORK_TEST_NET_CHAIN_ID,
        data: { blockchain, chainId }
    };
};

export const toggleTestNet = () => (dispatch: Dispatch<any>, getState: () => IReduxState) => {
    dispatch({
        type: SET_TEST_NET
    });
    const state = getState();

    const selectedBlockchain = getSelectedBlockchain(state);
    const networkExists = hasNetwork(Blockchain[selectedBlockchain], state.preferences.testNet);

    if (!networkExists) {
        const blockchains = getBlockchains(state);
        setSelectedBlockchain(Blockchain[blockchains[0]])(dispatch, getState);
    }

    const selectedAccount = getSelectedAccount(getState());
    getBalance(
        selectedAccount.blockchain,
        selectedAccount.address,
        undefined,
        true
    )(dispatch, getState);
};

export function toggleTouchID() {
    return {
        type: TOGGLE_TOUCH_ID
    };
}

export function setCurrency(currency: string) {
    return {
        type: PREF_SET_CURRENCY,
        data: { currency }
    };
}
