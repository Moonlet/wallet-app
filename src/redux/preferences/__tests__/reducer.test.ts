import reducer from '../reducer';
import * as actions from '../actions';
import { Blockchain } from '../../../core/blockchain/types';

describe('preferences reducer', () => {
    test('should set initial state', () => {
        expect(
            reducer(undefined as any, {
                type: '',
                data: ''
            })
        ).toMatchSnapshot();
    });

    test('should handle TOGGLE_PIN_LOGIN', () => {
        expect(
            reducer(undefined as any, {
                type: actions.TOGGLE_PIN_LOGIN,
                data: ''
            })
        ).toEqual({
            currency: 'USD',
            testNet: false,
            pinLogin: false,
            touchID: false,
            networks: {
                [Blockchain.ETHEREUM]: {
                    testNet: 4,
                    mainNet: 1
                },
                [Blockchain.ZILLIQA]: {
                    testNet: 333,
                    mainNet: 1
                }
            },
            blockchains: {
                [Blockchain.ETHEREUM]: {
                    order: 0,
                    active: true
                },
                [Blockchain.ZILLIQA]: {
                    order: 1,
                    active: true
                }
            }
        });
    });
});
