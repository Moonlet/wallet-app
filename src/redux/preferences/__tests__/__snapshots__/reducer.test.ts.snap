// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preferences reducer should set initial state 1`] = `
Object {
  "blockchains": Object {
    "ETHEREUM": Object {
      "active": true,
      "order": 1,
    },
    "NEAR": Object {
      "active": true,
      "order": 2,
    },
    "ZILLIQA": Object {
      "active": true,
      "order": 0,
    },
  },
  "currency": "USD",
  "networks": Object {
    "ETHEREUM": Object {
      "mainNet": 1,
      "testNet": 4,
    },
    "NEAR": Object {
      "testNet": "testnet",
    },
    "ZILLIQA": Object {
      "mainNet": 1,
      "testNet": 333,
    },
  },
  "testNet": false,
  "touchID": false,
}
`;
